In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
data=pd.read_csv(r'C:\Users\ssc\Desktop\Hackers_Rank\Training Data.csv')

In [3]:
data.head()

patient_id  Age Gender  Prescription_period  Diabetes  Alcoholism  \
0           1   19      M                    7         0           0   
1           2   24      F                   59         0           0   
2           3    4      F                   43         0           0   
3           4   38      M                   66         0           0   
4           5   46      F                   98         0           0   

   HyperTension  Smokes  Tuberculosis  Sms_Reminder Adherence  
0             0       0             0             0        No  
1             0       0             0             0        No  
2             0       0             0             0        No  
3             0       0             0             1        No  
4             0       0             0             1        No

In [4]:
print(data.columns)
print(data.shape)

Index(['patient_id', 'Age', 'Gender', 'Prescription_period', 'Diabetes',
       'Alcoholism', 'HyperTension', 'Smokes', 'Tuberculosis', 'Sms_Reminder',
       'Adherence'],
      dtype='object')
(180212, 11)


In [5]:
Model_Data=data[['Age','Prescription_period','Diabetes','Alcoholism','HyperTension','Smokes','Tuberculosis','Sms_Reminder','Adherence']]

In [6]:
Model_Data.head()

Age  Prescription_period  Diabetes  Alcoholism  HyperTension  Smokes  \
0   19                    7         0           0             0       0   
1   24                   59         0           0             0       0   
2    4                   43         0           0             0       0   
3   38                   66         0           0             0       0   
4   46                   98         0           0             0       0   

   Tuberculosis  Sms_Reminder Adherence  
0             0             0        No  
1             0             0        No  
2             0             0        No  
3             0             1        No  
4             0             1        No

In [7]:
Model_Data['Adherence'] = Model_Data['Adherence'].str.lower().replace({'yes': 1, 'no': 0})

In [8]:
Model_Data.head()

Age  Prescription_period  Diabetes  Alcoholism  HyperTension  Smokes  \
0   19                    7         0           0             0       0   
1   24                   59         0           0             0       0   
2    4                   43         0           0             0       0   
3   38                   66         0           0             0       0   
4   46                   98         0           0             0       0   

   Tuberculosis  Sms_Reminder  Adherence  
0             0             0          0  
1             0             0          0  
2             0             0          0  
3             0             1          0  
4             0             1          0

In [9]:
Model_Data.dropna(inplace=True)
Model_Data.describe()

Age  Prescription_period       Diabetes     Alcoholism  \
count  180212.000000        180212.000000  180212.000000  180212.000000   
mean       37.795363            54.668485       0.078524       0.025043   
std        22.852072            35.752491       0.268995       0.156255   
min         0.000000             1.000000       0.000000       0.000000   
25%        19.000000            22.000000       0.000000       0.000000   
50%        38.000000            51.000000       0.000000       0.000000   
75%        56.000000            86.000000       0.000000       0.000000   
max       113.000000           120.000000       1.000000       1.000000   

        HyperTension         Smokes   Tuberculosis   Sms_Reminder  \
count  180212.000000  180212.000000  180212.000000  180212.000000   
mean        0.216512       0.052566       0.000338       0.573968   
std         0.411868       0.223166       0.018395       0.499824   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       1.000000   
75%         0.000000       0.000000       0.000000       1.000000   
max         1.000000       1.000000       1.000000       2.000000   

           Adherence  
count  180212.000000  
mean        0.301811  
std         0.459045  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [10]:
all_features = Model_Data[['Age', 'Prescription_period',
                             'Diabetes', 'Alcoholism','HyperTension','Smokes','Tuberculosis','Sms_Reminder']].values


all_classes = Model_Data['Adherence'].values

feature_names = ['age', 'shape', 'margin', 'density']

all_features

array([[ 19,   7,   0, ...,   0,   0,   0],
       [ 24,  59,   0, ...,   0,   0,   0],
       [  4,  43,   0, ...,   0,   0,   0],
       ...,
       [ 48,  22,   0, ...,   0,   0,   0],
       [  5, 104,   0, ...,   0,   0,   0],
       [  0,  96,   0, ...,   0,   0,   1]], dtype=int64)

In [11]:
from sklearn import preprocessing

scaler = preprocessing.StandardScaler()
all_features_scaled = scaler.fit_transform(all_features)
all_features_scaled

array([[-0.8224819 , -1.33329499, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-0.60368275,  0.12115316, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.47887934, -0.32636935, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       ...,
       [ 0.44655317, -0.91374264, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.43511952,  1.37981021, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.65391866,  1.15604895, -0.29191724, ..., -0.2355469 ,
        -0.01840122,  0.85236588]])

# Decision Tree And Random Forest

In [12]:
import numpy
from sklearn.model_selection import train_test_split

numpy.random.seed(1234)

(training_inputs,
 testing_inputs,
 training_classes,
 testing_classes) = train_test_split(all_features_scaled, all_classes, train_size=0.75, random_state=1)

In [46]:
from sklearn.tree import DecisionTreeClassifier

clf= DecisionTreeClassifier(random_state=1)

# Train the classifier on the training set
clf.fit(training_inputs, training_classes)
y_pred=clf.predict(testing_inputs)

In [ ]:
Test_Model['Adherence']=clf.predict(all_features_scaled_Test)

In [27]:
Test_Model['Adherence'].head()

0    1
1    1
2    0
3    0
4    1
Name: Adherence, dtype: int64

In [14]:
 clf.score(testing_inputs, testing_classes)

0.8683994406587796

In [34]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,confusion_matrix

In [47]:
accuracy_score(testing_classes,y_pred)

0.8683994406587796

In [48]:
precision_score(testing_classes,y_pred)

0.7735234510712218

In [49]:
recall_score(testing_classes,y_pred)

0.7923926744272262

In [50]:
confusion_matrix(testing_classes,y_pred)

array([[28437,  3129],
       [ 2800, 10687]], dtype=int64)

In [15]:
from sklearn.model_selection import cross_val_score

clf = DecisionTreeClassifier(random_state=1)

cv_scores = cross_val_score(clf, all_features_scaled, all_classes, cv=10)

cv_scores.mean()

0.8701362941523401

In [16]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(n_estimators=10, random_state=1)
cv_scores = cross_val_score(clf, all_features_scaled, all_classes, cv=10)

cv_scores.mean()

0.8763956021886967

# Test File For Prediction

In [17]:
Test=pd.read_csv(r'C:\Users\ssc\Desktop\Hackers_Rank\Test Data.csv')

In [18]:
Test.head()

patient_id  Age Gender  Prescription_period  Diabetes  Alcoholism  \
0           1    5      M                   28         0           0   
1           2   62      F                    9         1           0   
2           3    4      F                   73         0           0   
3           4   33      M                  117         0           0   
4           5   38      M                    8         0           0   

   HyperTension  Smokes  Tuberculosis  Sms_Reminder  
0             0       0             0             1  
1             1       0             0             0  
2             0       0             0             1  
3             0       0             0             0  
4             0       0             0             1

In [19]:
Test_Model=Test[['Age','Prescription_period','Diabetes','Alcoholism','HyperTension','Smokes','Tuberculosis','Sms_Reminder']]

In [20]:
Test_Model.head()

Age  Prescription_period  Diabetes  Alcoholism  HyperTension  Smokes  \
0    5                   28         0           0             0       0   
1   62                    9         1           0             1       0   
2    4                   73         0           0             0       0   
3   33                  117         0           0             0       0   
4   38                    8         0           0             0       0   

   Tuberculosis  Sms_Reminder  
0             0             1  
1             0             0  
2             0             1  
3             0             0  
4             0             1

Predict Adherence with the Decision Tree

In [23]:
all_features_Test = Test_Model[['Age', 'Prescription_period',
                             'Diabetes', 'Alcoholism','HyperTension','Smokes','Tuberculosis','Sms_Reminder']].values
all_features_Test

array([[ 5, 28,  0, ...,  0,  0,  1],
       [62,  9,  1, ...,  0,  0,  0],
       [ 4, 73,  0, ...,  0,  0,  1],
       ...,
       [46, 95,  0, ...,  0,  0,  1],
       [27, 76,  0, ...,  0,  0,  1],
       [67, 89,  1, ...,  0,  0,  1]], dtype=int64)

In [24]:
scaler = preprocessing.StandardScaler()
all_features_scaled_Test = scaler.fit_transform(all_features_Test)
all_features_scaled

array([[-0.8224819 , -1.33329499, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-0.60368275,  0.12115316, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.47887934, -0.32636935, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       ...,
       [ 0.44655317, -0.91374264, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.43511952,  1.37981021, -0.29191724, ..., -0.2355469 ,
        -0.01840122, -1.14834476],
       [-1.65391866,  1.15604895, -0.29191724, ..., -0.2355469 ,
        -0.01840122,  0.85236588]])

# Predict the output 

In [31]:
# Train the classifier on the training set
clf.fit(training_inputs, training_classes)
Test_Model['Adherence']=clf.predict(all_features_scaled_Test)

Test_Model['Adherence'] = Test_Model['Adherence'].map({1: 'yes', 0: 'no'})

In [32]:
Test_Model['Adherence'].head()

0    yes
1    yes
2     no
3     no
4    yes
Name: Adherence, dtype: object

# Write Output File For upload

In [33]:
Test_Model.to_csv(r'C:\Users\ssc\Desktop\Hackers_Rank\Output.csv')

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix

In [ ]:
from sklearn.metrics import accuracy_score,precision_score,recall_score

# SVM

In [ ]:
from sklearn import svm

C = 1.0
svc = svm.SVC(kernel='linear', C=C)

In [ ]:
cv_scores = cross_val_score(svc, all_features_scaled, all_classes, cv=10)

cv_scores.mean()

# KNN

In [ ]:
from sklearn import neighbors

clf = neighbors.KNeighborsClassifier(n_neighbors=10)
cv_scores = cross_val_score(clf, all_features_scaled, all_classes, cv=10)

cv_scores.mean()

In [ ]:
for n in range(1, 50):
    clf = neighbors.KNeighborsClassifier(n_neighbors=n)
    cv_scores = cross_val_score(clf, all_features_scaled, all_classes, cv=10)
    print (n, cv_scores.mean())

# Naive Bayes

In [ ]:
from sklearn.naive_bayes import MultinomialNB

scaler = preprocessing.MinMaxScaler()
all_features_minmax = scaler.fit_transform(all_features)

clf = MultinomialNB()
cv_scores = cross_val_score(clf, all_features_minmax, all_classes, cv=10)

cv_scores.mean()

# Revisiting SVM

In [ ]:
C = 1.0
svc = svm.SVC(kernel='rbf', C=C)
cv_scores = cross_val_score(svc, all_features_scaled, all_classes, cv=10)
cv_scores.mean()

In [ ]:
C = 1.0
svc = svm.SVC(kernel='sigmoid', C=C)
cv_scores = cross_val_score(svc, all_features_scaled, all_classes, cv=10)
cv_scores.mean()

In [ ]:
C = 1.0
svc = svm.SVC(kernel='poly', C=C)
cv_scores = cross_val_score(svc, all_features_scaled, all_classes, cv=10)
cv_scores.mean()

# Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression()
cv_scores = cross_val_score(clf, all_features_scaled, all_classes, cv=10)
cv_scores.mean()

# Neural Network

In [ ]:
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential

def create_model():
    model = Sequential()
    #4 feature inputs going into an 6-unit layer (more does not seem to help - in fact you can go down to 4)
    model.add(Dense(6, input_dim=4, kernel_initializer='normal', activation='relu'))
    # "Deep learning" turns out to be unnecessary - this additional hidden layer doesn't help either.
    #model.add(Dense(4, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (benign or malignant)
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model; adam seemed to work best
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [ ]:
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=create_model, epochs=100, verbose=0)
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features_scaled, all_classes, cv=10)
cv_scores.mean()

# Decision Tree Gives Best Accuracy Amongs All Thats Why I Use Decision Tree For The  Prediction. we get 87% Accuracy for the Prediction which can be further improve by hyperparametric tunning.   